In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
features = pd.read_csv('/content/drive/MyDrive/Features data set.csv')
sales = pd.read_csv('/content/drive/MyDrive/sales data-set.csv')
stores = pd.read_csv('/content/drive/MyDrive/stores data-set.csv')

# Walmart Sales Analysis — Store Performance

## Store-Level Sales Analysis

Analyzing store performance using total sales, average sales, sales variability, and consistency.

## 1. Store Sales Performance

### Business Question:
Which stores generate the highest total sales?

In [ ]:
store_summary = sales.groupby('Store')['Weekly_Sales'].sum().reset_index()
print(store_summary.sort_values(by='Weekly_Sales',ascending=False).set_index('Store').head())

       Weekly_Sales
Store              
20     3.013978e+08
4      2.995440e+08
14     2.889999e+08
13     2.865177e+08
2      2.753824e+08


**Insight:** Store 20 has the highest total sales, followed by Stores 4, 14, 13, and 2.

### Business Question:
Which stores have the lowest total sales?

In [ ]:
store_summary = sales.groupby('Store')['Weekly_Sales'].sum().reset_index()
print(store_summary.sort_values(by='Weekly_Sales',ascending=True).set_index('Store').head())

       Weekly_Sales
Store              
33      37160221.96
44      43293087.84
5       45475688.90
36      53412214.97
38      55159626.42


**Insight:** Store 33 has the lowest total sales, while Stores 44, 5, 36, and 38 follow.

### Business Question:
Which store has the highest single-week sales?

In [22]:
sales.loc[sales['Weekly_Sales'].idxmax(),['Store','Date','Weekly_Sales']]

,95373
Store,10
Date,26/11/2010
Weekly_Sales,693099.36


**Insight:** Store 10 recorded the highest single-week sales of 693,099.36 on 26/11/2010.

### Business Question:
Which store had the lowest single-week sales?

In [23]:
sales.loc[sales['Weekly_Sales'].idxmin(),['Store','Date','Weekly_Sales']]

,267730
Store,28
Date,08/10/2010
Weekly_Sales,-4988.94


**Insight:** Store 28 recorded the lowest weekly sales of -4,988.94 on 08/10/2010.

### Business Question:
How do holiday and non-holiday weeks compare in total sales?

In [24]:
holiday_sales = sales.groupby('IsHoliday')['Weekly_Sales'].sum().reset_index()
holiday_sales.set_index('IsHoliday')

,Weekly_Sales
IsHoliday,
False,6.231919e+09
True,5.052996e+08


**Insight:** Non-holiday weeks generated substantially higher total sales than holiday weeks in the dataset.

### Business Question:
What is the average weekly sales for holiday vs. non-holiday weeks?

In [25]:
holiday_avg_sales = sales.groupby('IsHoliday')['Weekly_Sales'].mean().reset_index()
holiday_avg_sales.set_index('IsHoliday')


,Weekly_Sales
IsHoliday,
False,15901.445069
True,17035.823187


**Insight:** Average weekly sales were higher during holiday weeks than non-holiday weeks.

### Business Question:
Which departments generate the highest total sales?

In [26]:
department_summary = sales.groupby('Dept')['Weekly_Sales'].sum().reset_index()
department_summary.sort_values(by='Weekly_Sales',ascending=False).set_index('Dept').head(5)

,Weekly_Sales
Dept,
92,4.839433e+08
95,4.493202e+08
38,3.931181e+08
72,3.057252e+08
90,2.910685e+08


**Insight:** Department 92 generates the highest total sales, followed by Departments 95 and 38.

### Business Question:
Which departments have the lowest total sales?

In [27]:
department_summary = sales.groupby('Dept')['Weekly_Sales'].sum().reset_index()
department_summary.sort_values(by='Weekly_Sales',ascending=True).set_index('Dept').head(5)

,Weekly_Sales
Dept,
47,-4962.93
43,14.32
39,177.98
78,1714.71
51,30572.83


**Insight:** Department 47 has the lowest total sales and is the only department with negative total sales.

### Business Question:
Average sales by department

In [28]:
department_average = sales.groupby('Dept')['Weekly_Sales'].mean().reset_index()
department_average.sort_values(by='Weekly_Sales',ascending=False).set_index('Dept').head()

,Weekly_Sales
Dept,
92,75204.870531
95,69824.423080
38,61090.619568
72,50566.515417
65,45441.706224


**Insight:** Department 92 has the highest average weekly sales, followed by Departments 95 and 38.

### Business Question:
How did Walmart's total sales change from year to year?

In [29]:
sales['Date'] = pd.to_datetime(sales['Date'],dayfirst=True)
sales['Year'] = sales['Date'].dt.year

In [31]:
yearly_sales = sales.groupby('Year')['Weekly_Sales'].sum().reset_index()
yearly_sales.set_index('Year')

,Weekly_Sales
Year,
2010,2.288886e+09
2011,2.448200e+09
2012,2.000133e+09


**Insight:** Total sales increased from 2010 to 2011, then declined in 2012, with 2011 recording the highest total sales.

### Business Question:
monthly sales trends

In [33]:
sales['Date'] = pd.to_datetime(sales['Date'],dayfirst = True)
sales['Month'] = sales['Date'].dt.month

In [34]:
monthly_sales = sales.groupby('Month')['Weekly_Sales'].sum().reset_index()
monthly_sales.set_index('Month')

,Weekly_Sales
Month,
1,3.325984e+08
2,5.687279e+08
3,5.927859e+08
4,6.468598e+08
5,5.571256e+08
6,6.226299e+08
7,6.500010e+08
8,6.130902e+08
9,5.787612e+08


**Insight:** July recorded the highest total sales, while January had the lowest. Sales were generally stronger from March through October.

### Business Question:
sales by department using average weekly sales.

In [35]:
dept_avg = sales.groupby('Dept')['Weekly_Sales'].mean().reset_index()
dept_avg.sort_values(by='Weekly_Sales',ascending = False).set_index('Dept').head()

,Weekly_Sales
Dept,
92,75204.870531
95,69824.423080
38,61090.619568
72,50566.515417
65,45441.706224


**Insight:** Department 92 has the highest average weekly sales, followed by Departments 95 and 38.

In [36]:
dept_avg = sales.groupby('Dept')['Weekly_Sales'].mean().reset_index()
dept_avg.sort_values(by='Weekly_Sales',ascending = False).set_index('Dept').tail()

,Weekly_Sales
Dept,
51,21.931729
39,11.123750
78,7.296638
43,1.193333
47,-7.682554


**Insight:** Department 47 has the lowest average weekly sales, with a negative average, followed by Departments 43, 78, 39, and 51.

### Business Question:
holiday vs. non-holiday sales by year.

In [37]:
holiday_year_sales = sales.groupby(['Year','IsHoliday'])['Weekly_Sales'].sum().reset_index()
holiday_year_sales.sort_values(by='Weekly_Sales',ascending=False).set_index('Year')

,IsHoliday,Weekly_Sales
Year,,
2011,False,2.241465e+09
2010,False,2.088662e+09
2012,False,1.901793e+09
2011,True,2.067355e+08
2010,True,2.002246e+08
2012,True,9.833947e+07


**Insight:** Non-holiday sales were higher than holiday sales in every year, while holiday sales declined sharply in 2012.

### Business Question:
store performance by year

In [38]:
store_yearly_sales = sales.groupby(['Store','Year'])['Weekly_Sales'].sum().reset_index()
store_yearly_sales.sort_values(by='Weekly_Sales',ascending=False).set_index('Store').head()

,Year,Weekly_Sales
Store,,
4,2011,1.110923e+08
20,2011,1.098370e+08
14,2011,1.060963e+08
14,2010,1.054622e+08
13,2011,1.045375e+08


**Insight:** Store 4 in 2011 recorded the highest store-year sales, followed by Store 20 in 2011 and Store 14 in 2011.

In [39]:
store_yearly_sales = sales.groupby(['Store','Year'])['Weekly_Sales'].sum().reset_index()
store_yearly_sales.sort_values(by='Weekly_Sales',ascending=True).set_index('Store').head()

,Year,Weekly_Sales
Store,,
33,2012,11435551.03
33,2010,12766834.26
33,2011,12957836.67
36,2012,13286470.53
44,2010,13607519.45


**Insight:** Store 33 in 2012 recorded the lowest store-year sales, followed by Store 33 in 2010 and 2011.

### Business Question:
Which store has the highest average weekly sales?

In [40]:
store_avg = sales.groupby('Store')['Weekly_Sales'].mean().reset_index()
store_avg.sort_values(by='Weekly_Sales',ascending = False).set_index('Store').head()

,Weekly_Sales
Store,
20,29508.301592
4,29161.210415
14,28784.851727
13,27355.136891
2,26898.070031


**Insight:** Store 20 has the highest average weekly sales, followed by Stores 4, 14, 13, and 2.

In [41]:
store_avg = sales.groupby('Store')['Weekly_Sales'].mean().reset_index()
store_avg.sort_values(by='Weekly_Sales',ascending = True).set_index('Store').head()

,Weekly_Sales
Store,
5,5053.415813
33,5728.414053
44,6038.929814
3,6373.033983
38,7492.478460


**Insight:** Store 20 has the highest average weekly sales, followed by Stores 4, 14, 13, and 2.

### Business Question:
sales volatility by store.

In [42]:
store_volatility = sales.groupby('Store')['Weekly_Sales'].agg(
    max_sales = 'max',
    avg_sales = 'mean'
).reset_index()

store_volatility['sales_gap'] = store_volatility['max_sales'] - store_volatility['avg_sales']
store_volatility.sort_values(by='sales_gap',ascending = False).set_index('Store').head()

,max_sales,avg_sales,sales_gap
Store,,,
10,693099.36,26332.303819,666767.056181
35,649770.18,13803.596986,635966.583014
14,474330.10,28784.851727,445545.248273
27,420586.57,24826.984536,395759.585464
20,422306.25,29508.301592,392797.948408


**Insight:** Store 10 has the largest gap between its maximum and average weekly sales, followed by Stores 35 and 14, indicating strong peak-week variation.

In [43]:
store_volatility = sales.groupby('Store')['Weekly_Sales'].agg(
    max_sales = 'max',
    avg_sales = 'mean'
).reset_index()

store_volatility['sales_gap'] = store_volatility['max_sales'] - store_volatility['avg_sales']
store_volatility.sort_values(by='sales_gap',ascending = True).set_index('Store').head()

,max_sales,avg_sales,sales_gap
Store,,,
33,49129.77,5728.414053,43401.355947
44,66629.98,6038.929814,60591.050186
36,78974.83,8584.412563,70390.417437
30,80457.01,8764.237719,71692.772281
5,93517.72,5053.415813,88464.304187


**Insight:** Store 33 has the smallest gap between maximum and average weekly sales, indicating the least variation among the stores shown.

### Business Question:
which stores have the highest and lowest sales variability using standard deviation

In [44]:
store_std = sales.groupby('Store')['Weekly_Sales'].std().reset_index()
store_std.sort_values(by='Weekly_Sales',ascending=False).set_index('Store').head()

,Weekly_Sales
Store,
14,36911.116744
20,35661.530620
4,34583.677814
13,33853.474638
2,33077.612059


**Insight:** Store 14 has the highest sales variability, followed by Stores 20, 4, 13, and 2.

In [45]:
store_std = sales.groupby('Store')['Weekly_Sales'].std().reset_index()
store_std.sort_values(by='Weekly_Sales',ascending=True).set_index('Store').head()

,Weekly_Sales
Store,
5,8068.221050
33,8973.688840
29,10219.308122
16,10491.369269
44,10609.468795


**Insight:** Store 5 has the lowest sales variability, followed by Stores 33, 29, 16, and 44, indicating relatively more consistent weekly sales.

### Business Question:
average sales and sales variability together

In [46]:
store_risk = sales.groupby('Store')['Weekly_Sales'].agg(
    avg_sales = 'mean',
    sales_std = 'std'
).reset_index()
store_risk.sort_values(by='sales_std',ascending=False).set_index('Store').head()

,avg_sales,sales_std
Store,,
14,28784.851727,36911.116744
20,29508.301592,35661.530620
4,29161.210415,34583.677814
13,27355.136891,33853.474638
2,26898.070031,33077.612059


**Insight:** Store 14 shows the highest sales variability among these stores, while Store 20 combines the highest average sales with high variability.

### Business Question:
stores with high average sales but relatively low variability

In [47]:
store_risk.query(
    'avg_sales > 20000 & sales_std < 30000'
).sort_values(by='sales_std',ascending = False).set_index('Store')

,avg_sales,sales_std
Store,,
27,24826.984536,29718.919095
1,21710.543621,27748.945511
39,21000.763562,25373.751512
19,20362.126734,24281.125431
6,21913.243624,23633.427075


**Insight:** Store 27 leads this group in average sales, while Stores 6 and 39 combine strong average sales with comparatively lower variability.

### Business Question:
variability relative to average sales

In [48]:
store_risk['cv'] = round((store_risk['sales_std'] / store_risk['avg_sales']),2)
store_risk.sort_values(by='cv',ascending = False).set_index('Store').head()

,avg_sales,sales_std,cv
Store,,,
3,6373.033983,14251.034807,2.24
44,6038.929814,10609.468795,1.76
38,7492.478460,12733.356273,1.70
30,8764.237719,14364.314622,1.64
5,5053.415813,8068.221050,1.60


**Insight:** Store 3 has the highest relative sales variability (CV), followed by Stores 44 and 38, indicating less consistent sales relative to their average levels.

In [49]:
store_risk['cv'] = round((store_risk['sales_std'] / store_risk['avg_sales']),2)
store_risk.sort_values(by='cv',ascending = True).set_index('Store').head()

,avg_sales,sales_std,cv
Store,,,
11,19276.762751,20391.897907,1.06
23,19776.180881,20897.076552,1.06
6,21913.243624,23633.427075,1.08
8,13133.014768,15132.069598,1.15
21,11283.435496,13030.320859,1.15


**Insight:** Stores 11 and 23 have the lowest relative sales variability, indicating the most consistent sales among the stores shown.

### Business Question:
stores with negative weekly sales.

In [50]:
negative_sales = sales[sales['Weekly_Sales']<0]
negative_sales[['Store','Dept','Date','Weekly_Sales']].set_index('Store').sort_values(by='Weekly_Sales',ascending=True).head()

,Dept,Date,Weekly_Sales
Store,,,
28,6,2010-10-08,-4988.94
35,47,2011-09-02,-3924.00
45,47,2010-02-26,-1750.00
16,47,2012-03-09,-1699.00
28,32,2011-03-25,-1321.48


**Insight:** Negative sales occur across multiple stores and departments, with Store 28 recording the largest negative value among the displayed records.

### Business Question:
Which stores have the most negative-sales records?

In [51]:
negative_sales_count = sales[sales['Weekly_Sales']<0].groupby('Store').size().reset_index(name = 'negative_sales_count')
negative_sales_count.sort_values(by='negative_sales_count',ascending=False).head().reset_index(drop=True)

,Store,negative_sales_count
0,35,124
1,18,52
2,10,50
3,17,49
4,15,45


**Insight:** Store 35 has the highest frequency of negative-sales records with 124, substantially above the other top stores. This store should be investigated for possible data-quality issues or unusual sales activity.


### Business Question:
Which stores have the highest and lowest average weekly sales?

In [52]:
store_avg = sales.groupby('Store')['Weekly_Sales'].mean().reset_index(name='store_avg')
store_avg.sort_values(by='store_avg',ascending=False).reset_index(drop=True).head()

,Store,store_avg
0,20,29508.301592
1,4,29161.210415
2,14,28784.851727
3,13,27355.136891
4,2,26898.070031


**Insight:** Store 20 has the highest average weekly sales at 29,508.30, followed by Stores 4 and 14. These stores demonstrate the strongest typical weekly sales performance among the top five.


In [53]:
store_avg = sales.groupby('Store')['Weekly_Sales'].mean().reset_index(name='store_avg')
store_avg.sort_values(by='store_avg',ascending=True).reset_index(drop=True).head()

,Store,store_avg
0,5,5053.415813
1,33,5728.414053
2,44,6038.929814
3,3,6373.033983
4,38,7492.478460


**Insight:** Store 5 has the lowest average weekly sales at 5,053.42, followed by Stores 33 and 44. These stores show the weakest typical weekly sales performance among the bottom five.


### Business Question:
Which stores have the highest variability in weekly sales?

In [54]:
store_std = sales.groupby('Store')['Weekly_Sales'].std().reset_index(name='store_std')
store_std.sort_values(by='store_std',ascending=False).reset_index(drop=True).head()

,Store,store_std
0,14,36911.116744
1,20,35661.530620
2,4,34583.677814
3,13,33853.474638
4,2,33077.612059


**Insight:** Store 14 has the highest weekly sales variability with a standard deviation of 36,911.12, followed by Stores 20 and 4. These stores experience greater week-to-week fluctuations in sales.


### Business Question:
Which stores have the lowest variability in weekly sales?

In [55]:
store_std = sales.groupby('Store')['Weekly_Sales'].std().reset_index(name='store_std')
store_std.sort_values(by='store_std',ascending=True).reset_index(drop=True).head()

,Store,store_std
0,5,8068.221050
1,33,8973.688840
2,29,10219.308122
3,16,10491.369269
4,44,10609.468795


**Insight:** Store 5 has the lowest weekly sales variability with a standard deviation of 8,068.22, followed by Stores 33 and 29. These stores show the most consistent weekly sales among the five.


### Business Question:
Which stores combine high average weekly sales with relatively low sales variability?

In [56]:
store_risk = store_avg.merge(store_std,on='Store')
store_risk.sort_values(by='store_std').set_index('Store').head(5)

,store_avg,store_std
Store,,
5,5053.415813,8068.221050
33,5728.414053,8973.688840
29,8158.810609,10219.308122
16,7863.224124,10491.369269
44,6038.929814,10609.468795


**Insight:** Among the five most consistent stores, Store 29 has the highest average weekly sales at 8,158.81, while Store 5 has the lowest variability at 8,068.22. This shows the importance of considering both sales level and consistency when evaluating store performance.


In [9]:
strong_stable_stores = store_risk.query(
    "store_avg >20000 & store_std <30000"
)
print(strong_stable_stores.sort_values(by='store_avg',ascending=False).reset_index(drop=True))

   Store     store_avg     store_std
0     27  24826.984536  29718.919095
1      6  21913.243624  23633.427075
2      1  21710.543621  27748.945511
3     39  21000.763562  25373.751512
4     19  20362.126734  24281.125431


**Insight:** Store 27 leads the strong-and-stable group with the highest average weekly sales, while Stores 6 and 39 combine strong sales with comparatively lower variability.


### Business Question:
Which stores have strong average sales with the lowest relative variability?

In [57]:
store_risk['cv'] = round((store_risk['store_std']/store_risk['store_avg']),2)
strong_stable_cv = store_risk.query("store_avg > 20000")
strong_stable_cv.sort_values(by="cv",ascending = True).set_index('Store').head()

,store_avg,store_std,cv
Store,,,
6,21913.243624,23633.427075,1.08
4,29161.210415,34583.677814,1.19
19,20362.126734,24281.125431,1.19
27,24826.984536,29718.919095,1.20
20,29508.301592,35661.530620,1.21


**Insight:** Store 6 has the lowest relative sales variability (CV 1.08), while Store 20 achieves the highest average sales but with greater relative fluctuation.
